proto_proxy.c

```diff
@@ -0,0 +1,2310 @@
+    lua_State *L = ctx;
+    // TODO: stash into a table with weak references?
+    // then if no pools/code has references still, can ditch?
+    // TODO: __gc
+    const struct luaL_Reg mcplib_backend_m[] = {
+        {"set", NULL},
+        {NULL, NULL}
+    };
+
+    const struct luaL_Reg mcplib_request_m[] = {
+        {"command", mcplib_request_command},
+        {"key", mcplib_request_key},
+        {"__tostring", NULL},
+        {"__gc", mcplib_request_gc},
+        {NULL, NULL}
+    };
+
+    const struct luaL_Reg mcplib_response_m[] = {
+        {"ok", mcplib_response_ok},
+        {"__gc", mcplib_response_gc},
+        {NULL, NULL}
+    };
+
+    // TODO: __gc
+    const struct luaL_Reg mcplib_hash_selector_m[] = {
+        {"__call", mcplib_hash_selector_call},
+        {NULL, NULL}
+    };
+
+    const struct luaL_Reg mcplib_f [] = {
+        {"hash_selector", mcplib_hash_selector},
+        {"backend", mcplib_backend},
+        {"request", mcplib_request},
+        {"attach", mcplib_attach},
+        {NULL, NULL}
+    };
+
+    // TODO: function + loop.
+    luaL_newmetatable(L, "mcp.backend");
+    lua_pushvalue(L, -1); // duplicate metatable.
+    lua_setfield(L, -2, "__index"); // mt.__index = mt
+    luaL_setfuncs(L, mcplib_backend_m, 0); // register methods
+    lua_pop(L, 1);
+
+    luaL_newmetatable(L, "mcp.request");
+    lua_pushvalue(L, -1); // duplicate metatable.
+    lua_setfield(L, -2, "__index"); // mt.__index = mt
+    luaL_setfuncs(L, mcplib_request_m, 0); // register methods
+    lua_pop(L, 1);
+
+    luaL_newmetatable(L, "mcp.response");
+    lua_pushvalue(L, -1); // duplicate metatable.
+    lua_setfield(L, -2, "__index"); // mt.__index = mt
+    luaL_setfuncs(L, mcplib_response_m, 0); // register methods
+    lua_pop(L, 1);
+
+    luaL_newmetatable(L, "mcp.hash_selector");
+    lua_pushvalue(L, -1); // duplicate metatable.
+    lua_setfield(L, -2, "__index"); // mt.__index = mt
+    luaL_setfuncs(L, mcplib_hash_selector_m, 0); // register methods
+    lua_pop(L, 1); // drop the hash selector metatable
+
+    // create main library table.
+    //luaL_newlib(L, mcplib_f);
+    // TODO: luaL_newlibtable() just pre-allocs the exact number of things
+    // here.
+    // can replace with createtable and add the num. of the constant
+    // definitions.
+    luaL_newlibtable(L, mcplib_f);
+    proxy_register_defines(L);
+
+    // hash function for selectors.
+    // have to wrap the function in a struct because function pointers aren't
+    // pointer pointers :)
+    mcplib_open_jump_hash(L);
+    lua_setfield(L, -2, "hash_jump");
+    // FIXME: remove this once multi-probe is in, use that as default instead.
+    lua_pushlightuserdata(L, &mcplib_hashfunc_murmur3);
+    lua_setfield(L, -2, "hash_murmur3");
+
+    lua_pushlightuserdata(L, (void *)t); // upvalue for original thread
+    lua_newtable(L); // upvalue for mcp.attach() table.
+
+    luaL_setfuncs(L, mcplib_f, 2); // 2 upvalues.
+
```

Comment:
What is the state of stack at this point?

Reply:
After line 2306 the stack should only have the mcp table in it. The setglobal line then pops it and assigns it to the `mcp` global value.